### DISCUSSION TITLE
Stale server side atom value.

### DISCUSSION THREAD

**Discussion by t1ger-0527** [pmndrs/jotai#1603] [Q&A]
In the [codesandbox example](https://jotai.org/docs/utils/use-hydrate-atoms#codesandbox) for `useHydrateAtoms` hook, the `lastUpdate` value consumed from `clockAtom` within the Clock component stays as the same value when I first visit the `/ssr` page. All subsequent `/ssr` page visit will result in the same time even though in `pages/ssr` the latest `Date.now()` is passed into initialState as the initial value.

experiment: https://codesandbox.io/s/nextjs-with-jotai-forked-4rdnv7

I added 2 `console.log`  on top of the the original example, 1 in `_app` to log down the initial value passed into the app, 2nd in `Clock` component to reveal the actual atom value used for render.

![screenshot of console output](https://user-images.githubusercontent.com/30690934/204587936-31d7e955-c7f1-4ac5-8f41-b5cb142537e4.png)

So my question is: what's the proper way to make sure the `clockAtom` to always get the latest value in each render session?

**Comment by dai-shi** [comment]
> the lastUpdate value consumed from clockAtom within the Clock component stays as the same value when I first visit the /ssr page.

Yeah, that's how it works. And this behavior is only safe for concurrent rendering.
It's close to how `useState(initialValue)` works. Though, useState is reset on remount.

> what's the proper way to make sure the clockAtom to always get the latest value in each render session?

We need to combine with useEffect if you want to keep track of the prop value.
```js
  const { initialState } = pageProps;
  useHydrateAtoms(initialState ? [[clockAtom, initialState]] : []);
  const setClock = useSetAtom(clockAtom);
  useEffect(() => {
    setClock(initialState);
  }, [setClock, initialState]);
```

**Reply by t1ger-0527** [reply]
After trying your suggestion a little bit, I have a feeling that you might have missed my point.

The issue here is that the `clockAtom` used on the server side always holds the first value that is set into the atom, instead of the updated one from `pageProps`. I believe the latest initial state passed into `pageProps` will also be used in hydration, and it caused the issue where:
- on the server side, `Clock` uses the stale value from `clockAtom` (that lives on the server side)
- on the client side, `Clock` uses the latest value passed from the server inside `pageProps` when `useHydrateAtoms` is called.

Although on the page behavior it looks OK when we see the correct value with dark background and it turns to light when it is hydrated. But if you take a look at the HTML response, as well as the console you will realize it's actually broken.

From the HTML response:
![image](https://user-images.githubusercontent.com/30690934/204748945-d8db0cf8-b8bd-4167-8bf8-4f58d9ee7ad9.png)

From the console:
![image](https://user-images.githubusercontent.com/30690934/204749055-5dde47bb-8557-4bac-85e2-1af41127fb76.png)

**Reply by t1ger-0527** [reply]
And as for your suggestion, I believe it won't help this issue because `useEffect` will not be executed on the server side.

**Reply by dai-shi** [reply]
> you might have missed my point.

sounds like so.

> The issue here is that the clockAtom used on the server side always holds the first value that is set into the atom

that sounds like you are using provider-less mode which is not good on server.
you should have `Provider` in _app.js. https://jotai.org/docs/guides/nextjs#provider

**Reply by t1ger-0527** [reply]
Thanks for your prompt reply! Let me try the Provider in the example.
